File: Python/normalize.py
import numpy as np

from math import sqrt
def line_to_int(liste):
    for i in range(0,len(liste)):
            liste[i]=int(liste[i])
pic="../img/img_1.ppm"
fichier=open(pic,"r")
lines = [line.rstrip('\n') for line in fichier]
matrix=[]
for line in range(3,len(lines)):
    raw=[]
    ligne=[]
    pixel=[]
    raw=lines[line].split()
    line_to_int(raw)
    for i in range(0,len(raw)):
        if(i%3==0):
            pixel+=[raw[i]]
        if(i%3==1):
            pixel+=[raw[i]]  
        if(i%3==2):
            pixel+=[raw[i]]  
            ligne+=[pixel]
            pixel=[]
    matrix+=[ligne]
    print(len(matrix))
#print(matrix)
fichier.close()
f=open("matrix.txt","w")
f.write("".join(map(str, matrix)))
f.close()
img=[]
for i in range(4,28):
    img+=[matrix[i]]
#for i in range(len(img)):
 #   del img[i][0:4]
  #  del img[i][28:31]
#print(len(img[20]))>img.txt

print(np.array(matrix))
for i in range(4):
    del matrix[0]
    del matrix[-1]
for i in matrix:
    for j in range(4):
        del i[0]
        del i[-1]
def normalise(matrix):
    N=24*24*3
    mu=0
    sigma=0
    for i in range(len(matrix)):
        for j in range(len(matrix)):
            for k in range(3):
                mu+=matrix[i][j][k]/N
    for i in range(len(matrix)):
        for j in range(len(matrix)):
            for k in range(3):
                sigma+=(matrix[i][j][k]-mu)**2
    sigma=sqrt(sigma/N)
    for i in range(len(matrix)):
        for j in range(len(matrix)):
            for k in range(3):
                matrix[i][j][k]=(matrix[i][j][k]-mu)/(max(sigma,1/sqrt(N)))
    return matrix
def create_fmap(image):
    # On convertit l'image en un tableau numpy
    image = np.array(image)
    # On transpose les dimensions de l'image pour passer de (10,10,3) à (3,10,10)
    image = np.transpose(image, (2, 0, 1))
    # On crée une fmap vide de la même taille que l'image transposée
    fmap = np.zeros_like(image)
    # On parcourt chaque canal de l'image
    for c in range(image.shape[0]):
        # On extrait chaque canal de l'image
        channel = image[c,:,:]
        # On effectue un traitement sur chaque canal, par exemple une convolution avec un filtre
        # ...
        # On stocke le résultat dans la fmap correspondante
        fmap[c,:,:] = channel
    return fmap
matrix=normalise(matrix)
matrix=create_fmap(matrix)
print(matrix)
fi=open("imag.txt","w")
fi.write("".join(map(str, np.array(matrix))))
fi.close()

File: C++/normalize.cpp
#include <fstream>
#include <iostream>
#include <cmath>
#include <vector>
#include <string>
#include <sstream>

using namespace std;

// Converts a string to an integer
double str_to_double(string str) {
    stringstream ss(str);
    double result;
    ss >> result;
    return result;
}

// Converts a list of strings to a list of integers
void str_list_to_double(vector<string> &list1,vector<double> &list) {
    for (int i = 0; i < list1.size(); i++) {
        list.push_back(str_to_double(list1[i]));
    }
}
void print_matrix(vector<vector<vector<double>>> matrix) {
    for (int i = 0; i < matrix.size(); i++) {
        for (int j = 0; j < matrix[i].size(); j++) {
            for (int k = 0; k < matrix[i][j].size(); k++) {
                cout << matrix[i][j][k] << " ";
            }
            cout << endl;
        }
        cout << endl;
    }
}

// ...




int main() {
    // Open pic.ppm file in read mode
    ifstream file("../img/img_1.ppm");

    // Read in all lines from the file
    vector<string> lines;
    string line;
    while (getline(file, line)) {
        lines.push_back(line);
    }
    file.close();

    // Convert lines to a matrix of integers
    vector<vector<vector<double>>> matrix;
    for (int i = 3; i < lines.size(); i++) {
        vector<vector<double>> ligne;
        vector<string> raw1;
        vector<double> raw;
        vector<double> pixel;
        stringstream ss(lines[i]);
        string token;
        while (getline(ss, token, ' ')) {
            raw1.push_back(token);
        }
        str_list_to_double(raw1,raw);
        for (int j = 0; j < raw.size(); j++) {
            if (j % 3 == 0) {
                pixel.push_back(raw[j]);
            }
            if (j % 3 == 1) {
                pixel.push_back(raw[j]);
            }
            if (j % 3 == 2) {
                pixel.push_back(raw[j]);
                ligne.push_back(pixel);
                pixel.clear();
            }
        }
        matrix.push_back(ligne);
        cout << matrix.size() << endl;
    }

    // Write matrix to matrix.txt file
    ofstream f("matrix.txt");
    for (int i = 0; i < matrix.size(); i++) {
        for (int j = 0; j < matrix[i].size(); j++) {
            for (int k = 0; k < matrix[i][j].size(); k++) {
                f << matrix[i][j][k] << " ";
            }
            f << endl;
        }
        f << endl;
    }
    f.close();

    // Create a new matrix with the rows 4 to 27
    vector<vector<vector<double>>> img;
    for (int i = 4; i < 28; i++) {
        vector<vector<double>> row;
        for (int j = 4; j < 28; j++) {
            row.push_back(matrix[i][j]);
        }
        img.push_back(row);
    }

    // Normalize the values in the img matrix
    int N = 24 * 24 * 3;
    double mu = 0;
    double sigma = 0;
    for (int i = 0; i < img.size(); i++) {
    for (int j = 0; j < img[i].size(); j++) {
        for (int k = 0; k < img[i][j].size(); k++) {
            mu += img[i][j][k] / N;
        }
    }
}

for (int i = 0; i < img.size(); i++) {
    for (int j = 0; j < img[i].size(); j++) {
        for (int k = 0; k < img[i][j].size(); k++) {
            sigma += (img[i][j][k] - mu) * (img[i][j][k] - mu);
        }
    }
}
sigma = sqrt(sigma / N);

for (int i = 0; i < img.size(); i++) {
    for (int j = 0; j < img[i].size(); j++) {
        for (int k = 0; k < img[i][j].size(); k++) {
            img[i][j][k] = (img[i][j][k] - mu) / max(sigma, 1 / sqrt(N));
        }
    }
}
print_matrix(img);
// Write img to img.txt file
ofstream f1("img.txt");
for (int i = 0; i < img.size(); i++) {
    for (int j = 0; j < img[i].size(); j++) {
        for (int k = 0; k < img[i][j].size(); k++) {
            f1 << img[i][j][k] << " ";
            }
        f1 << endl;
    }
    f1 << endl;
}
f1.close();
return 0;

}